[clang][dataflow] Factor out built-in boolean model into an explicit module.

ymand opened the pull request:
Draft to demo how we can pull out the boolean model. Let's discuss specifics of
namings, location, etc.

The purpose of this refactoring is to enable us to compare the performance of
different boolean models. In particular, we're interested in investigating a
very simple semantic domain of just the booleans (and Top).

In the process, the PR drastically simplifies the handling of terminators. This
cleanup can be pulled out into its own PR, to precede the refactoring work.

ymand committed:
[clang][dataflow] Factor out built-in boolean model into an explicit module.

In the process, drastically simplify the handling of terminators.

martinboehme commented:
> Draft to demo how we can pull out the boolean model. Let's discuss specifics of namings, location, etc.

Not sure -- do you mean let's wordsmith names now, or do you mean we should discuss naming and location, but that should happen after we've talked about the general approach?

> The purpose of this refactoring is to enable us to compare the performance of different boolean models. In particular, we're interested in investigating a very simple semantic domain of just the booleans (and Top).

Can you expand on how we would swap in a different boolean model?

*  Just put `#ifdef`s in the various functions in `bool_model`?

*  Provide different namespaces containing different boolean models, then in `namespace bool_model`, do `using namespace my_desired_bool_model`?

*  Something else?

I would favour a model that's as simple as possible -- I don't think we want to use template parameters, for example -- and what you have here looks like it's intended to be simple. I'm just not sure exactly where this is intended to go?

> In the process, the PR drastically simplifies the handling of terminators. This cleanup can be pulled out into its own PR, to precede the refactoring work.

I like the cleanup, and I think pulling it out into a separate patch is a good idea because a) it's unrelated to the rest of this patch, and b) it can land today, without further discussion needed (IMO).

ymand commented:
> > Draft to demo how we can pull out the boolean model. Let's discuss specifics of namings, location, etc.
> 
> Not sure -- do you mean let's wordsmith names now, or do you mean we should discuss naming and location, but that should happen after we've talked about the general approach?

Either way -- I just meant that I'm not tied to the particulars included in the draft.

> > The purpose of this refactoring is to enable us to compare the performance of different boolean models. In particular, we're interested in investigating a very simple semantic domain of just the booleans (and Top).
> 
> Can you expand on how we would swap in a different boolean model?
> 
> * Just put `#ifdef`s in the various functions in `bool_model`?
> * Provide different namespaces containing different boolean models, then in `namespace bool_model`, do `using namespace my_desired_bool_model`?
> * Something else?

For starters, the namespace approach. That will give us a simple way to experiment with alternatives. But, next step is to turn this namespace into a derivative of `DataflowModel` -- we'll just need to extend that interface to support `transferBranch`.

Additional alternatives:

1. template parameter -- make the boolean model a static parameter of the overall system. This sounds like a nightmare, to be blunt, forcing a huge amount of code into templates and massive rewriting. Let's not.
2. multiple models -- the boolean model actions occur inside functions that are already parameterized. If we would support multiple models, then we could simply remove the boolean modeling altogether from the core and bundle it up as a standalone model. That is, the core will not concern itself with boolean modeling. I like this for the long term, but don't want to block on this for now.

> 
> I would favour a model that's as simple as possible -- I don't think we want to use template parameters, for example -- and what you have here looks like it's intended to be simple. I'm just not sure exactly where this is intended to go?
> 
> > In the process, the PR drastically simplifies the handling of terminators. This cleanup can be pulled out into its own PR, to precede the refactoring work.
> 
> I like the cleanup, and I think pulling it out into a separate patch is a good idea because a) it's unrelated to the rest of this patch, and b) it can land today, without further discussion needed (IMO).

Will do!

ymand commented:
Terminator cleanup split out into https://github.com/llvm/llvm-project/pull/84499

ymand committed:
[clang][dataflow] Add implementation of a simple boolean model

The new model still uses atomic variables in boolean formulas, but it limits the
environment to accumulating truth values for atomic variables, rather than
the arbitrary formula allowed by the flow condition.

martinboehme commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -1059,9 +1066,16 @@ void Environment::assume(const Formula &F) {
+#else
+bool Environment::proves(const Formula &F) const {
+  auto V = simple_bool_model::getLiteralValue(F, *this);
+  return V.has_value() && *V;
```
```suggestion
  return V.value_or(false);
```

martinboehme commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -1059,9 +1066,16 @@ void Environment::assume(const Formula &F) {
+#endif
 
 bool Environment::allows(const Formula &F) const {
   return DACtx->flowConditionAllows(FlowConditionToken, F);
```
Wouldn't this also need to be changed? I _think_ this can just do `return proves(F);`?

ymand replied:
Yes, but not to `proves`: it should be `value_or(true)` -- that is, if we lack any definite setting, we "allow" it to be anything.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Transfer.cpp:
```diff
@@ -50,29 +50,206 @@ const Environment *StmtToEnvMap::getEnvironment(const Stmt &S) const {
+    return Env.getAtomValue(F.getAtom());
+  case Formula::Literal:
+    return F.literal();
+  case Formula::Not: {
```
Why do we need this case? Is it not covered by `neOp()` above? (I.e. are there ever any cases where we actually have `Formula::Not` formulas?)

If we do need this, don't we also need corresponding cases for `Formula::And` and `Formula::Or`?

ymand replied:
Sorry, this got clobbered when I pushed after a rebase. But, this was on the function `getLiteralValue` (now in DataflowEnvironment.cpp) and I don't think the original concerns apply, since we now support And and Or, etc.

ymand committed:
[clang][dataflow] Abstract the logical operations (assume, allows, proves) and drop distinction between boolean operations.

This commit drastically simplifies the original refactoring. We keep the boolean
model separately, but we only maintain one version, since there turned out to be
no meaningful difference between them. Instead, the difference lies in the
logical operations, so we've abstacted those.

We're down to 35 failing tests, all with clear explanations based on the
limitations of this approach; primarily, the inability to encode custom
API/operator meanings using logical formulae.

ymand commented:
Martin, I've thoroughly updated the refactoring, exactly as you suggested -- all of the interesting differences are actually just in how we handle the logical operations, so most of the changes are now in DataflowEnvironment.cpp.  I've left the factoring in Transfer because we may want it for the future -- these two models both use formulae, but other implementations could differ. 

The test failures are down to 35 and all they are all WAI -- places where we have genuine differences between the models, primarily around encoding custom API properties with formulae.
